vmm: Add vm.add-fs route

deansheather opened the pull request:
Adds an API method for creating vhost-user-fs devices after the VM is launched.

This is mostly just copypasta'd from `vm.add-disk`.

deansheather commented:
I'll squash after this gets reviewed.

sboeuf commented on docs/api.md:
```diff
@@ -93,8 +93,9 @@ Dump the VM information            | `/vm.info`          | N/A
 Add VFIO PCI device to the VM      | `/vm.add-device`    | `/schemas/VmAddDevice`    | N/A               | The VM is booted
 Remove VFIO PCI device from the VM | `/vm.remove-device` | `/schemas/VmRemoveDevice` | N/A               | The VM is booted
 Add disk device to the VM          | `/vm.add-disk`      | `/schemas/DiskConfig`     | N/A               | The VM is booted
+Add FS device to the VM            | `/vm.add-fs`        | `/schemas/FsConfig`       | N/A               | The VM is booted
```
Nit: s/`FS`/`fs/`

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -1325,88 +1328,91 @@ impl DeviceManager {
+        let fs_devices = self.config.lock().unwrap().fs.clone();
+        if let Some(fs_list_cfg) = &fs_devices {
+            for fs_cfg in fs_list_cfg.iter() {
+                if let Some(_) = fs_cfg.sock.to_str() {
```
You don't need this line as it is checked from inside `make_virtio_fs_device()` directly.

sboeuf commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -187,6 +187,22 @@ paths:
+        required: true
+      responses:
+        204:
+          description: The new FS was successfully added to the VM instance.
```
The new device was successfully added to the VM instance.

sboeuf commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -187,6 +187,22 @@ paths:
+        204:
+          description: The new FS was successfully added to the VM instance.
+        500:
+          description: The new FS could not be added to the VM instance.
```
The new device could not be added to the VM instance.

sboeuf commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -187,6 +187,22 @@ paths:
 
+  /vm.add-fs:
+    put:
+      summary: Add a new virtio-fs backed FS device to the VM
```
Simply `Add a new virtio-fs device to the VM` is enough.

sboeuf commented on vmm/src/config.rs:
```diff
@@ -864,6 +864,11 @@ impl Default for FsConfig {
 }
 
 impl FsConfig {
+    pub const SYNTAX: &'static str = "FS parameters \
```
Could you please copy/paste from what is already in `src/main.rs`?
And don't forget to reuse this new `SYNTAX` from `src/main.rs` too.

sboeuf commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -187,6 +187,22 @@ paths:
+    put:
+      summary: Add a new virtio-fs backed FS device to the VM
+      requestBody:
+        description: The details of the new FS
```
Nit: s/`FS`/`virtio-fs/`

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -1325,88 +1328,91 @@ impl DeviceManager {
-                    let migratable = Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn Migratable>>;
-                    let id = migratable.lock().unwrap().id();
-                    self.migratable_devices.insert(id, migratable);
+        Err(DeviceManagerError::NoVirtioFsSock)
```
Thanks for adding this error, this would prevent the code from silently ignoring malformed `--fs` parameters.

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -1411,7 +1411,7 @@ impl DeviceManager {
         if let Some(fs_list_cfg) = &fs_devices {
             for fs_cfg in fs_list_cfg.iter() {
-                if let Some(_) = fs_cfg.sock.to_str() {
+                if fs_cfg.sock.to_str().is_some() {
```
This is not needed as `make_virtio_fs_device()` performs a check for this already.

deansheather commented:
First commit creates an API endpoint that just returns `501 Not Implemented`. Second commit makes the required changes to device_manager.rs and implements the API.

sboeuf commented:
@rbradford PTAL

rbradford commented:
@deansheather Thanks for the PR. What's the use case that you have in mind that requires hotplugging a filesystem? (We didn't include it in the original set of hotpluggable devices as we didn't have a clear motivator.)

deansheather commented:
@rbradford Coder is building a CRI runtime based on cloud-hypervisor and we need the ability to add virtio-fs devices (for container bind mounts to the host) after a VM is started.

rbradford commented:
> @rbradford Coder is building a CRI runtime based on cloud-hypervisor and we need the ability to add virtio-fs devices (for container bind mounts to the host) after a VM is started.

Okay cool! Please add an integration test (see `tests/integration.rs` `test_disk_hotplug` and `test_virtio_fs` for inspiration.) When the test is ready we should be able to proceed with final review and merge.

deansheather commented:
I can't get hotplugged dax devices to work in my integration tests, which is strange because hotplugging dax devices seems to work externally. I'm not a vmm guy so I have no idea what's causing the below error:

```
cloud-hypervisor: 20.016993704s: ERROR:pci/src/bus.rs:242 -- Failed moving device BAR: bus_error: MissingAddressRange
```

This error happens when the add-fs call is made to the API when the fs has dax enabled.

rbradford commented:
> I can't get hotplugged dax devices to work in my integration tests, which is strange because hotplugging dax devices seems to work externally. I'm not a vmm guy so I have no idea what's causing the below error:
> 
> ```
> cloud-hypervisor: 20.016993704s: ERROR:pci/src/bus.rs:242 -- Failed moving device BAR: bus_error: MissingAddressRange
> ```
> 
> This error happens when the add-fs call is made to the API when the fs has dax enabled.

Might be a longshot but try incorporating #1027 into your PR.

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -775,16 +775,11 @@ impl Vm {
+                // device automatically gets recreated on reboot. Unfoutunately,
+                // virtiofsd exits when the fuse client disconnects, which means
+                // automatically recreating the device will cause the reboot to
+                // fail.
```
The problem you're seeing with virtio-fs and reboot is not tied to the hotplug, it's simply not supported to reboot with a virtio-fs device attached.
That's why the problem needs to be fixed by having virtio-fs self spawning being supported at some point.
So I'd suggest that this PR should keep the initial code you wrote, as this is semantically correct.

sboeuf commented on tests/integration.rs:
```diff
@@ -1912,6 +1912,122 @@ mod tests {
+            Ok(())
+        });
+    }
+
```
You need to edit the commit title and commit message to indicate you're adding integration tests.

deansheather replied:
The tests fail, so I pushed it as a temporary commit. Once the issue is fixed I'll amend the commit.

sboeuf replied:
Ok so about the failing test, is it working locally but not with the CI?

Have you tried to rebase on latest master as suggested by @rbradford since master now includes #1027 ?

deansheather replied:
Fail locally, the ci didn't run for some reason so I can't check. I tried again with the #1027 patch before it was merged, I'll try again with a rebase on master.

sboeuf replied:
Oh it fails locally, then this is gonna be easier to debug. Let me know if you have made some progress on it, otherwise I'll try to help with it :)

deansheather commented:
Still doesn't work after a rebase on master unfortunately. Not sure how to proceed in fixing this since I don't have much experience with vmm programming.

```
cloud-hypervisor: 20.022188101s: ERROR:pci/src/bus.rs:242 -- Failed moving device BAR: bus_error: MissingAddressRange: 0x3ffdffe00000->0x200000000(0x200000000)
```

sboeuf commented:
@deansheather 
>Still doesn't work after a rebase on master unfortunately. Not sure how to proceed in fixing this since I don't have much experience with vmm programming.

Ok I'll have a look at it tomorrow. Could you rebase your PR on latest master and push it? It'll be easier for me to start from there :)

sboeuf commented:
@deansheather just wanted to give you a heads up that the problem is a bit trickier than expected, but I have a plan to fully fix this :)
In the meantime, you can cleanup your PR and make it ready for when I come up with the code.
You can even test your work by adding the two commits I added to your branch on my personal branch `sboeuf/fix_add-fs`.
My branch is not complete but that's enough to get the virtio-fs hotplug working, so that should be enough to let you test and finalize your code.

sboeuf commented:
@deansheather ok so I have some good news, your PR is now passing the CI.
I have rebased it on top of #1041, so it can't be merged until #1041 is merged.
One more thing, I addressed the comment I left in your code, and I also added one more commit to factorize the integration test into `test_virtio_fs`, preventing from too much duplication.
Please have a look and let me know if something's wrong :)

deansheather commented:
@sboeuf This looks great! Thanks for fixing this and addressing the comments for me, sorry for not being available to sort those out myself over the weekend.

sboeuf commented:
@deansheather 
>This looks great! Thanks for fixing this and addressing the comments for me, sorry for not being available to sort those out myself over the weekend.

No worries :)

Can you check that it works correctly on your side, so that you can start using the feature as soon as we merge it?

Also, it'd be nice for you to send the PR for hot-unplugging `virtio-fs` device when you get a chance!

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -1414,93 +1417,97 @@ impl DeviceManager {
+        }
+    }
+
+    fn make_virtio_fs_devices(
```
Have you tested creating multiple fs devices?

sboeuf replied:
Actually no, I can have a try, but right now I don't know why the CI is not passing...

deansheather replied:
I'm seeing the same issues in our project, the second virtio-fs device won't mount and the guest kernel reports that the tag was not found.

sboeuf replied:
Ok the PR is fixed! I added one more patch (a oneliner changing quite a lot of things) to make everything work. I'm glad I factorized the integration test because it made me realize we had a problem as virtio-fs was not working after some memory was hotplugged.

About multiple devices, I've just tested that and it works fine for me. You need to run multiple backends and then hotplug them one after the other. If you mount them in different locations on the host it should work for you too :)

deansheather replied:
Seems to be working but I'm hitting a connection limit wall even after bumping the connection limit constant up... I'm trying to add 10 or so FSes, and the first 8 or so seem to work

sboeuf replied:
What do you mean by connection limit?

I'm curious why you need so many virtio-fs devices to be passed into a single VM. Are you passing one virtio-fs device per container that you run into the VM?
I'm asking because you can bypass this limitation by having one single virtio-fs device shared across all containers in the same VM.

One more thing, I never tried so many `virtio-fs` devices, but there's a hard limit set at 32 PCI devices since we emulate only one PCI bus with Cloud-Hypervisor.

deansheather replied:
That makes sense, we'll just stick to one virtio-fs then per VM.
